src/GovUk.Education.ExploreEducationStatistics.Publisher/Services/DataSetPublishingService.cs

```diff
@@ -44,77 +42,88 @@ private async Task PromoteDraftDataSetVersions(IEnumerable<Guid> releaseVersionI
-                // set.
-                dataSet.Status = DataSetStatus.Published;
-                dataSet.Published ??= currentDraftVersion.Published;
-                
-                // Set the newly published DataSetVersion as the overarching DataSet's LatestLiveVersion,
-                // and unset it from being the latest Draft version.
-                dataSet.LatestLiveVersion = currentDraftVersion;
-                dataSet.LatestLiveVersionId = currentDraftVersion.Id;
-                dataSet.LatestDraftVersion = null;
-                dataSet.LatestDraftVersionId = null;
-                
-                publicDataDbContext.DataSets.Update(dataSet);
-                publicDataDbContext.DataSetVersions.Update(currentDraftVersion);
-            });
-        
-        publicDataDbContext.DataSets.UpdateRange(releaseVersionDataSets);
+        publicDataDbContext.DataSets.UpdateRange(dataSets);
+
+        foreach (var dataSet in dataSets)
+        {
+            var currentDraftVersion = dataSet.LatestDraftVersion;
+            currentDraftVersion!.Status = DataSetVersionStatus.Published;
+            currentDraftVersion.Published = DateTimeOffset.UtcNow;
+
+            dataSet.Status = DataSetStatus.Published;
+            dataSet.Published ??= currentDraftVersion.Published;
+
+            dataSet.LatestLiveVersion = currentDraftVersion;
+            dataSet.LatestLiveVersionId = currentDraftVersion.Id;
+
+            dataSet.LatestDraftVersion = null;
+            dataSet.LatestDraftVersionId = null;
+
+            publicDataDbContext.DataSetVersions.Update(currentDraftVersion);
+        }
 
         await publicDataDbContext.SaveChangesAsync();
     }
 
     /// <summary>
-    /// This method updates any ReleaseFileId values on DataSetVersions that referenced any Ids of ReleaseFiles on
-    /// previous ReleaseVersions, in the case that any of the given releaseVersionIds that are being published are
-    /// amendments of ReleaseVersions that contain Public API Data sets.
+    /// For any amendment release, update any relevant data set versions being published
+    /// with IDs that reference the amendment's files.
     /// </summary>
-    private async Task UpdateReleaseFileIdsForReleaseAmendment(IEnumerable<Guid> releaseVersionIds)
+    private async Task UpdateDataSetVersionFileIdsForAmendments(IEnumerable<Guid> releaseVersionIds)
     {
         var previousReleaseVersionIds = await contentDbContext
             .ReleaseVersions
-            .Where(releaseVersion =>
-                releaseVersionIds.Contains(releaseVersion.Id)
-                && releaseVersion.PreviousVersionId != null)
+            .Where(releaseVersion => releaseVersionIds.Contains(releaseVersion.Id))
+            .Where(releaseVersion => releaseVersion.PreviousVersionId != null)
             .Select(releaseVersion => releaseVersion.PreviousVersionId)
             .Cast<Guid>()
             .ToListAsync();
```

Comment:
could be a `HashSet`

Reply:
In this instance, I've just changed it by adding `Distinct`. The ids are just being supplied to another LINQ query so the extra wrangling into a HashSet isn't really necessary.